File: backend/src/main/java/vn/duytan/c1se09/smartsurvey/service/NotificationService.java
package vn.duytan.c1se09.smartsurvey.service;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import vn.duytan.c1se09.smartsurvey.domain.Notification;
import vn.duytan.c1se09.smartsurvey.domain.User;
import vn.duytan.c1se09.smartsurvey.repository.NotificationRepository;

import java.util.List;

/**
 * Service để quản lý notifications
 */
@Service
@RequiredArgsConstructor
@SuppressWarnings("null")
public class NotificationService {

    private final NotificationRepository notificationRepository;

    /**
     * Tạo notification mới
     */
    @Transactional
    public Notification createNotification(User user, Notification.NotificationType type, 
                                          String title, String message, 
                                          String relatedEntityType, Long relatedEntityId) {
        Notification notification = new Notification();
        notification.setUser(user);
        notification.setType(type);
        notification.setTitle(title);
        notification.setMessage(message);
        notification.setRelatedEntityType(relatedEntityType);
        notification.setRelatedEntityId(relatedEntityId);
        notification.setIsRead(false);
        return notificationRepository.save(notification);
    }

    /**
     * Lấy danh sách notifications của user
     */
    public List<Notification> getUserNotifications(User user) {
        return notificationRepository.findByUserOrderByCreatedAtDesc(user);
    }

    /**
     * Lấy danh sách notifications chưa đọc
     */
    public List<Notification> getUnreadNotifications(User user) {
        return notificationRepository.findByUserAndIsReadFalseOrderByCreatedAtDesc(user);
    }

    /**
     * Đếm số notifications chưa đọc
     */
    public long getUnreadCount(User user) {
        return notificationRepository.countByUserAndIsReadFalse(user);
    }

    /**
     * Đánh dấu notification là đã đọc
     */
    @Transactional
    public void markAsRead(Long notificationId, User user) {
        Notification notification = notificationRepository.findById(notificationId)
                .orElseThrow(() -> new RuntimeException("Không tìm thấy notification"));
        
        if (!notification.getUser().getUserId().equals(user.getUserId())) {
            throw new RuntimeException("Bạn không có quyền đánh dấu notification này");
        }
        
        notification.setIsRead(true);
        notificationRepository.save(notification);
    }

    /**
     * Đánh dấu tất cả notifications là đã đọc
     */
    @Transactional
    public void markAllAsRead(User user) {
        notificationRepository.markAllAsRead(user);
    }
}

File: frontend/src/services/notificationService.js
import { apiClient } from './authService';

export const notificationService = {
    /**
     * GET /api/notifications
     * Lấy danh sách notifications của user hiện tại
     */
    getNotifications: async () => {
        try {
            const response = await apiClient.get('/api/notifications');
            return response.data;
        } catch (error) {
            console.error('Get notifications error:', error);
            throw error;
        }
    },

    /**
     * GET /api/notifications/unread
     * Lấy danh sách notifications chưa đọc
     */
    getUnreadNotifications: async () => {
        try {
            const response = await apiClient.get('/api/notifications/unread');
            return response.data;
        } catch (error) {
            console.error('Get unread notifications error:', error);
            throw error;
        }
    },

    /**
     * GET /api/notifications/unread/count
     * Đếm số notifications chưa đọc
     */
    getUnreadCount: async () => {
        try {
            const response = await apiClient.get('/api/notifications/unread/count');
            return response.data;
        } catch (error) {
            console.error('Get unread count error:', error);
            throw error;
        }
    },

    /**
     * PUT /api/notifications/{notificationId}/read
     * Đánh dấu notification là đã đọc
     */
    markAsRead: async (notificationId) => {
        try {
            const response = await apiClient.put(`/api/notifications/${notificationId}/read`);
            return response.data;
        } catch (error) {
            console.error('Mark notification as read error:', error);
            throw error;
        }
    },

    /**
     * PUT /api/notifications/read-all
     * Đánh dấu tất cả notifications là đã đọc
     */
    markAllAsRead: async () => {
        try {
            const response = await apiClient.put('/api/notifications/read-all');
            return response.data;
        } catch (error) {
            console.error('Mark all notifications as read error:', error);
            throw error;
        }
    }
};

export default notificationService;

